server/irmaserver/helpers.go

```diff
@@ -498,74 +546,96 @@ func (s *Server) cacheMiddleware(next http.Handler) http.Handler {
 	})
 }
 
-func (s *Server) sessionMiddleware(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		token := chi.URLParam(r, "token")
-		session, err := s.sessions.clientGet(token)
-		if err != nil {
-			server.WriteError(w, server.ErrorInternal, "")
-			return
-		}
-		if session == nil {
-			server.WriteError(w, server.ErrorSessionUnknown, "")
-			return
-		}
+func (s *Server) sessionMiddleware(readOnly []string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			token := irma.ClientToken(chi.URLParam(r, "clientToken"))
+			session, err := s.sessions.clientGet(token)
+			if err != nil {
+				server.WriteError(w, server.ErrorInternal, "")
+				return
+			}
+			if session == nil {
+				server.WriteError(w, server.ErrorSessionUnknown, "")
+				return
+			}
 
-		ctx := r.Context()
-		session.Lock()
-		session.locked = true
-		defer func() {
-			if session.PrevStatus != session.Status {
-				session.PrevStatus = session.Status
-				result := session.Result
-				r := ctx.Value("sessionresult")
-				if r != nil {
-					*r.(*server.SessionResult) = *result
-				}
-				if session.Status.Finished() {
-					if handler := s.handlers[result.Token]; handler != nil {
-						go handler(result)
-						delete(s.handlers, result.Token)
-					}
+			ctx := r.Context()
+
+			// by default session needs to be locked
+			lock := true
+
+			// for certain endpoints read access for Redis without locking is sufficient
+			for _, e := range readOnly {
+				if strings.HasSuffix(r.URL.Path, e) && s.conf.StoreType == "redis" {
+					lock = false
 				}
 			}
-			if session.locked {
-				session.locked = false
-				session.Unlock()
+
+			// lock the session
+			if lock {
+				err = session.sessions.lock(session)
+				if err != nil {
+					server.WriteError(w, server.ErrorInternal, "")
+					return
+				}
```

Comment:
How does the locking work for functionality being called via the Go API? Because when for example the `CancelSession` API function is called, the `sessionMiddleware` is not run.

Reply:
In these cases, the lock method is called directly as there are only a couple of cases. See e.g. in the servers `CancelSession` method.

Reply:
Not sure what you mean by that. If I recall correctly the session status calls are get calls only. With each get the latest session version is fetched from the session store, just then without locking the session which is not necessary cause it's read-only. Do I miss something here?